packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts

```diff
@@ -404,63 +493,261 @@ const handler: AdaptingProxyHandler<ProxyTarget, EditableTree> = {
+                    configurable: true,
+                    enumerable: true,
+                    value: target.proxifyNode(Number(key)),
+                    writable: false,
+                };
+            }
+        }
+        return undefined;
+    },
+};
+
 /**
  * See {@link UnwrappedEditableTree} for documentation on what unwrapping this performs.
  */
-function inProxyOrUnwrap(target: ProxyTarget, unwrap: boolean): UnwrappedEditableTree {
-    if (unwrap) {
-        const fieldSchema = target.getType(undefined, false) as NamedTreeSchema;
-        if (isPrimitive(fieldSchema)) {
+function inProxyOrUnwrap(
+    target: NodeProxyTarget | FieldProxyTarget,
+    unwrap: boolean,
+): UnwrappedEditableTree {
+    // Unwrap primitives or nodes having a primary field. Sequences unwrap nodes on their own.
+    if (unwrap && !isFieldProxyTarget(target)) {
+        const nodeType = target.getType(undefined, false) as NamedTreeSchema;
+        if (isPrimitive(nodeType)) {
             const nodeValue = target.cursor.value;
             if (isPrimitiveValue(nodeValue)) {
                 return nodeValue;
             }
             assert(
-                fieldSchema.value === ValueSchema.Serializable,
+                nodeType.value === ValueSchema.Serializable,
                 0x3c7 /* `undefined` values not allowed for primitive fields */,
             );
         }
         const primary = target.getPrimaryArrayKey();
         if (primary !== undefined) {
             target.cursor.enterField(primary.key);
-            const childTargets = mapCursorField(
+            const primarySequence = new FieldProxyTarget(
+                target.context,
+                primary.schema,
                 target.cursor,
-                (c) => new ProxyTarget(target.context, c),
+                nodeType.name,
             );
-            target.cursor.exitField();
-            return childTargets.map((childTarget) => inProxyOrUnwrap(childTarget, unwrap));
+            target.free();
+            return adaptWithProxy(primarySequence, fieldProxyHandler);
         }
     }
-    return adaptWithProxy(target, handler);
+    return isFieldProxyTarget(target)
+        ? adaptWithProxy(target, fieldProxyHandler)
+        : adaptWithProxy(target, nodeProxyHandler);
 }
 
 /**
+ * @param context - the common context of the field.
  * @param fieldSchema - the FieldSchema of the field.
  * @param fieldKey - the key of the field. Used to visualize the tree.
- * @param childTargets - targets for the children of the field.
- * @param unwrap - if true, the children of the field are unwrapped (see {@link UnwrappedEditableField}),
- * otherwise returns the field as {@link EditableField}.
+ * @param cursor - the cursor, which must point to the field being proxified.
+ * @param unwrap - if true, the field is unwrapped (see {@link UnwrappedEditableField}),
+ * otherwise always returns the field as {@link EditableField}.
  */
 export function proxifyField(
```

Comment:
since this is not part of the package API, it should probably take the ProxyContext type directly.

if we do want to leave it taking just EditableTreeContext, it should prabably assert its is a ProxyContext (and do sone once instead of at every use, ex `assert(context instanceof ProxyContext, "requires ProxyContext");` at the top.

Reply:
Done.